src/Swarm/Language/LSP/VarUsage.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Swarm.Language.LSP.VarUsage where
+
+import Control.Monad (guard)
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Text (Text)
+import Data.Text qualified as T
+import Swarm.Language.Parse qualified as P
+import Swarm.Language.Syntax
+
+data DownwardDecl
+  = Lambda
+  | Let
+
+-- TODO: Unused-checking for these is not yet implemented
+data UpwardDecl
+  = Def
+  | Bind
+
+data VarScope
+  = Downward DownwardDecl
+  | Upward UpwardDecl
+
+data VarUsage = VarUsage Syntax VarScope Var
+
+-- | NOTE: We want existing keys to be overridden as we descend deeper
+-- into the tree, since inner scopes have precedence.
+type DownScopes = Map Var (DownwardDecl, Syntax)
+
+type UpScopes = Map Var [(UpwardDecl, Syntax)]
+
+-- | This positioning is customized per language construct.
+-- E.g. the "Syntax" node for @let@ includes the literal string \"let\"
+-- as well as an equals sign.
+--
+-- TODO: Perhaps if the language parser represented
+-- the "Var" as a \"located\" "Syntax" node as well, we would not need to
+-- reverse-engineer bits of the grammar here.
```

Comment:
This should definitely be possible.  I'm surprised the parser doesn't already do that.

Reply:
Yes, let's replace the `Var` in these constructors with a `LocVar` type.  Will you do so in another PR?

It's almost tempting to parameterize the `Syntax` type over what is now the `term` field, so we can stick `Var` there instead.

Reply:
Sure, I'll see about making another PR with `LocVar` filled in by the parser.